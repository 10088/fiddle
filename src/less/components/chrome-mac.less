.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 23px;
    font-size: 1.4rem;

    small,
    p,
    span {
      user-select: none;
    }
  }
}
