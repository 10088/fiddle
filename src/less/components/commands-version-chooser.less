button.version-chooser.bp3-button {
  border-radius: 3px;
}
