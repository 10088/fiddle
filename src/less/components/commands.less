@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-1;
  -webkit-app-region: drag;
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding: 10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      button {
        margin-left: 5px;
      }
    }
  }

  .address-bar {
    height: 28px;
    width: 360px;
    transition: width 0.4s;
    display: flex;

    input,
    span {
      font-family: @fonts-common;
      font-weight: 300;
      font-size: 11px;
      cursor: auto;
      -webkit-app-region: no-drag;
    }

    input {
      flex-grow: 1;
      background: white;
      padding-left: 1rem;
      height: 2.6rem;
    }

    input:invalid {
      box-shadow: inset 0 0 0 1px #e4252d;
    }

    span {
      background: white;
      color: #1d1d1d;
      border-bottom-color: rgb(209, 209, 209);
      border-bottom-style: solid;
      border-bottom-left-radius: 4px;
      border-bottom-width: 1px;
      border-image-outset: 0;
      border-image-repeat: stretch;
      border-image-slice: 100%;
      border-image-source: none;
      border-image-width: 1;
      border-left-color: rgb(209, 209, 209);
      border-left-style: solid;
      border-left-width: 1px;
      border-top-color: rgb(209, 209, 209);
      border-top-left-radius: 4px;
      border-top-style: solid;
      border-top-width: 1px;
      display: inline-block;
      height: 1.2rem;
      letter-spacing: normal;
      margin: 0;
      padding: 6px 0 10px 10px;
      user-select: text;
      width: 132px;
      text-rendering: auto;
      text-shadow: none;
      text-transform: none;
    }

    &.empty {
      width: 200px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }
}
