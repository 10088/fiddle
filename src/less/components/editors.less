.mosaic-tile {
  animation: fadein 0.15s;
}

.mosaic-window-toolbar  {
  & > div {
    display: flex;
    justify-content: space-between;
    width: 100%;
    padding: 0 3px 0 29px;

    .mosaic-controls {
      transform: scale(0.7);
    }

    h5 {
      margin: 4px 0 0 0;
    }

    button {
      margin-left: 8px;
    }
  }
}

@keyframes fadein {
  from { opacity: 0; }
  to { opacity: 1; }
}
