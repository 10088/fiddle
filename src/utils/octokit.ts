import * as GitHubType from '@octokit/rest';

let _Octokit: typeof GitHubType;
let _octo: GitHubType;

/**
 * Returns a loaded Octokit.
 *
 * @export
 * @returns {Promise<typeof GitHubType>}
 */
export async function getOctokit(): Promise<GitHubType> {
  _Octokit = _Octokit || await import('@octokit/rest');
  return _octo = _octo || new _Octokit();
}
