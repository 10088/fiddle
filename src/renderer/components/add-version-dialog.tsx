import { observer } from 'mobx-react';
import * as path from 'path';
import * as React from 'react';
import * as semver from 'semver';

import { NpmVersion } from '../../interfaces';
import { getElectronNameForPlatform } from '../../utils/electron-name';
import { AppState } from '../state';
import { Dialog } from './dialog';

export interface AddVersionDialogProps {
  appState: AppState;
}

export interface AddVersionDialogState {
  isValidElectron: boolean;
  isValidVersion: boolean;
  file?: File;
  version: string;
}

/**
 * The "add version" dialog allows users to add custom builds of Electron.
 *
 * @class AddVersionDialog
 * @extends {React.Component<AddVersionDialogProps, AddVersionDialogState>}
 */
@observer
export class AddVersionDialog extends React.Component<AddVersionDialogProps, AddVersionDialogState> {
  constructor(props: AddVersionDialogProps) {
    super(props);

    this.state = {
      isValidVersion: false,
      isValidElectron: false,
      version: ''
    };

    this.onSubmit = this.onSubmit.bind(this);
    this.onClose = this.onClose.bind(this);
    this.onChangeFile = this.onChangeFile.bind(this);
    this.onChangeVersion = this.onChangeVersion.bind(this);
  }

  /**
   * Handles a change of the file input
   *
   * @param {React.ChangeEvent<HTMLInputElement>} event
   */
  public async onChangeFile(event: React.ChangeEvent<HTMLInputElement>) {
    const { files } = event.target;
    const { binaryManager } = this.props.appState;
    const file = files && files[0] ? files[0] : undefined;

    const isValidElectron = !!(file
      && file.path
      && await binaryManager.getIsDownloaded('custom', file.path));

    this.setState({ file, isValidElectron });
  }

  public onChangeVersion(event: React.ChangeEvent<HTMLInputElement>) {
    const version = event.target.value || '';
    const isValidVersion = !!semver.valid(version);

    this.setState({
      version,
      isValidVersion,
     });
  }

  /**
   * Handles the submission of the dialog
   *
   * @returns {Promise<void>}
   */
  public async onSubmit(): Promise<void> {
    const { file, version } = this.state;

    if (!file) return;

    const name = file.path
      .slice(-20)
      .split(path.sep)
      .slice(1)
      .join(path.sep);

    const toAdd: NpmVersion = {
      localPath: file.path,
      version,
      name
    };

    this.props.appState.addLocalVersion(toAdd);
    this.onClose();
  }

  /**
   * Closes the dialog
   */
  public onClose() {
    this.props.appState.isAddVersionDialogShowing = false;
    this.reset();
  }

  /**
   * Reset this component's state
   */
  public reset(): void {
    this.setState({
      isValidElectron: false,
      isValidVersion: false,
      version: '',
      file: undefined
    });
  }

  get buttons() {
    const canSubmit = this.state.isValidElectron && this.state.isValidVersion;

    return [
      (
        <button
          className='button'
          key='submit'
          disabled={!canSubmit}
          onClick={this.onSubmit}
        >
          Add
        </button>
      ), (
        <button
          className='cancel'
          key='cancel'
          onClick={this.onClose}
        >
          Cancel
        </button>
      )
    ];
  }

  public render() {
    const { isAddVersionDialogShowing } = this.props.appState;
    const dirOptions = { webkitdirectory: 'true' };

    return (
      <Dialog
        isShowing={isAddVersionDialogShowing}
        isShowingBackdrop={true}
        buttons={this.buttons}
        onClose={this.onClose}
        isCentered={true}
        className='add-version-dialog'
        key='add-version-dialog'
      >
        <label
          htmlFor='custom-electron-version'
          className='force-button'
        >
          Select the folder containing {getElectronNameForPlatform()}
        </label>
        <input
          type='file'
          onChange={this.onChangeFile}
          id='custom-electron-version'
          name='custom-electron-version'
          {...dirOptions}
        />
        {this.renderPath()}
      </Dialog>
    );
  }

  private renderPath(): JSX.Element | null {
    const { file, isValidElectron } = this.state;

    if (!file || !file.path) return null;

    const info = isValidElectron
      ? `We found an ${getElectronNameForPlatform()} in this folder.`
      : `We did not find a ${getElectronNameForPlatform()} in this folder...`;

    return (
      <>
        <input
          readOnly={true}
          value={file.path}
        />
        <p>
          {info}
        </p>
        {this.renderVersionInput()}
      </>
    );
  }

  private renderVersionInput(): JSX.Element | null {
    const { isValidElectron, isValidVersion, version } = this.state;

    if (!isValidElectron) return null;

    return (
      <>
        <p>
          Please specify a version, used for typings and the name.
          Must be <code>semver</code> compliant.
        </p>
        <input
          className={isValidVersion ? '' : 'hasError'}
          value={version}
          onChange={this.onChangeVersion}
          placeholder='4.0.0'
        />
      </>
    );
  }

}
