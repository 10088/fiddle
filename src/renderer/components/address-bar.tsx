import * as React from 'react';
import { observer } from 'mobx-react';
import * as Octokit from '@octokit/rest';
import * as classNames from 'classnames';

import { AppState } from '../state';
import { INDEX_HTML_NAME, MAIN_JS_NAME, RENDERER_JS_NAME } from '../constants';
import { idFromUrl } from '../../utils/gist';
import { reaction } from 'mobx';

export interface AddressBarProps {
  appState: AppState;
}

export interface AddressBarState {
  value: string;
}

@observer
export class AddressBar extends React.Component<AddressBarProps, AddressBarState> {
  constructor(props: AddressBarProps) {
    super(props);

    this.loadFiddle = this.loadFiddle.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);

    this.state = {
      value: this.props.appState.gistId
    };
  }

  /**
   * Handle the form's submit event, trying to load whatever
   * URL was entered.
   *
   * @param {React.SyntheticEvent<HTMLFormElement>} event
   * @memberof AddressBar
   */
  public async handleSubmit(event: React.SyntheticEvent<HTMLFormElement>) {
    event.preventDefault();

    this.props.appState.gistId = idFromUrl(this.state.value) || this.state.value;
    if (this.state.value) {
      this.loadFiddle();
    }
  }

  /**
   * Once the component mounts, we'll subscribe to gistId changes
   */
  public componentDidMount() {
    reaction(
      () => this.props.appState.gistId,
      (gistId: string) => this.setState({ value: gistId })
    );
  }

  /**
   * Handle the change event, which usually just updates the address bar's value
   *
   * @param {React.ChangeEvent<HTMLInputElement>} event
   * @memberof AddressBar
   */
  public handleChange(event: React.ChangeEvent<HTMLInputElement>) {
    this.setState({ value: idFromUrl(event.target.value) || event.target.value });
  }

  /**
   * Load a fiddle
   *
   * @returns
   * @memberof AddressBar
   */
  public async loadFiddle() {
    const { gistId } = this.props.appState;

    if (!confirm('Are you sure you want to load a new fiddle, all current progress will be lost?')) return;

    try {
      const octo = new Octokit();
      const gist = await octo.gists.get({
        gist_id: gistId,
        id: gistId,
      });

      window.ElectronFiddle.app.setValues({
        html: gist.data.files[INDEX_HTML_NAME].content,
        main: gist.data.files[MAIN_JS_NAME].content,
        renderer: gist.data.files[RENDERER_JS_NAME].content,
      });
    } catch (error) {
      console.warn(`Loading fiddle failed`, error);
    }
  }

  public render() {
    const { isUnsaved } = this.props.appState;
    const { value } = this.state;
    const className = classNames('address-bar', isUnsaved, { empty: !value });

    return (
      <form className={className} onSubmit={this.handleSubmit}>
        <span>https://gist.github.com/</span>
        <input
          key='addressbar'
          type='text'
          placeholder='...'
          value={value}
          onChange={this.handleChange}
        />
      </form>
    );
  }
}
