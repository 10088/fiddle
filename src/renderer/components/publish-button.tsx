import { faSpinner, faUpload } from '@fortawesome/fontawesome-free-solid';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { observer } from 'mobx-react';
import * as React from 'react';

import { when } from 'mobx';
import { INDEX_HTML_NAME, MAIN_JS_NAME, RENDERER_JS_NAME } from '../../constants';
import { IpcEvents } from '../../ipc-events';
import { classNames } from '../../utils/classnames';
import { getOctokit } from '../../utils/octokit';
import { ipcRendererManager } from '../ipc';
import { AppState } from '../state';

export interface PublishButtonProps {
  appState: AppState;
}

export interface PublishButtonState {
  isPublishing: boolean;
}

/**
 * The "publish" button takes care of logging you in.
 *
 * @export
 * @class PublishButton
 * @extends {React.Component<PublishButtonProps, PublishButtonState>}
 */
@observer
export class PublishButton extends React.Component<PublishButtonProps, PublishButtonState> {
  constructor(props: PublishButtonProps) {
    super(props);

    this.state = { isPublishing: false };
    this.handleClick = this.handleClick.bind(this);
    this.publishFiddle = this.publishFiddle.bind(this);
  }

  public componentDidMount() {
    ipcRendererManager.on(IpcEvents.FS_SAVE_FIDDLE_GIST, this.handleClick);
  }

  /**
   * When the user clicks the publish button, we either show the
   * authentication dialog or publish right away.
   *
   * If we're showing the authentication dialog, we wait for it
   * to be closed again (or a GitHub token to show up) before
   * we publish
   *
   * @returns {Promise<void>}
   * @memberof PublishButton
   */
  public async handleClick(): Promise<void> {
    const { appState } = this.props;

    if (!appState.gitHubToken) {
      appState.toggleAuthDialog();
    }

    // Wait for the dialog to be closed again
    await when(() => !!appState.gitHubToken || !appState.isTokenDialogShowing);

    if (appState.gitHubToken) {
      return this.publishFiddle();
    }
  }

  /**
   * Connect with GitHub, publish the current Fiddle as a gist,
   * and update all related properties in the app state.
   */
  public async publishFiddle(): Promise<void> {
    this.setState({ isPublishing: true });

    const octo = await getOctokit();
    octo.authenticate({
      type: 'token',
      token: this.props.appState.gitHubToken!
    });

    const options = { includeDependencies: true, includeElectron: true };
    const values = await window.ElectronFiddle.app.getValues(options);


    try {
      const gist = await octo.gists.create({
        public: true,
        description: 'Electron Fiddle Gist',
        files: {
          [INDEX_HTML_NAME]: {
            content: values.html || '<!-- Empty -->',
          },
          [MAIN_JS_NAME]: {
            content: values.main || '// Empty',
          },
          [RENDERER_JS_NAME]: {
            content: values.renderer || '// Empty',
          },
        },
      } as any); // Note: GitHub messed up, GistsCreateParamsFiles is an incorrect interface

      this.props.appState.gistId = gist.data.id;

      console.log(`Publish Button: Publishing done`, { gist });
    } catch (error) {
      console.warn(`Could not publish gist`, { error });

      const messageBoxOptions: Electron.MessageBoxOptions = {
        message: 'Publishing Fiddle to GitHub failed. Are you connected to the Internet?',
        detail: `GitHub encountered the following error: ${error.message}`
      };

      ipcRendererManager.send(IpcEvents.SHOW_WARNING_DIALOG, messageBoxOptions);
    }

    this.setState({ isPublishing: false });
  }

  public render() {
    const { isPublishing } = this.state;
    const className = classNames('button', 'button-publish', isPublishing);
    const icon = isPublishing ? faSpinner : faUpload;
    const text = isPublishing ? 'Publishing...' : 'Publish';

    return (
      <button className={className} onClick={this.handleClick} disabled={isPublishing}>
        <FontAwesomeIcon icon={icon} spin={isPublishing} />
        <span style={{ marginLeft: 8 }} />
        {text}
      </button>
    );
  }
}
