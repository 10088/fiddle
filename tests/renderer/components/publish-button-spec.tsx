import * as React from 'react';
import { shallow } from 'enzyme';

import { PublishButton } from '../../../src/renderer/components/publish-button';
import { getOctokit } from '../../../src/utils/octokit';

jest.mock('../../../src/utils/octokit');
jest.mock('electron', () => require('../../mocks/electron'));

describe('Publish button component', () => {
  beforeEach(() => {
    this.store = {};
  });

  it('renders', () => {
    const wrapper = shallow(<PublishButton appState={this.store} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('toggles the auth dialog on click if not authed', () => {
    this.store.toggleAuthDialog = jest.fn();

    const wrapper = shallow(<PublishButton appState={this.store} />);
    wrapper.find('button').simulate('click');
    expect(this.store.toggleAuthDialog).toHaveBeenCalled();
  });

  it('toggles the publish method on click if authed', async () => {
    this.store.gitHubToken = 'github-token';

    const wrapper = shallow(<PublishButton appState={this.store} />);
    const instance: PublishButton = wrapper.instance() as any;
    instance.publishFiddle = jest.fn();
    await instance.handleClick();

    expect(instance.publishFiddle).toHaveBeenCalled();
  });

  it('attempts to publish to Gist', async () => {
    const mockOctokit = {
      authenticate: jest.fn(),
      gists: {
        create: jest.fn(async () => ({ data: { id: '123' } }))
      }
    };

    (getOctokit as any).mockReturnValue(mockOctokit);

    const wrapper = shallow(<PublishButton appState={this.store} />);
    const instance: PublishButton = wrapper.instance() as any;

    await instance.publishFiddle();

    expect(mockOctokit.authenticate).toHaveBeenCalled();
    expect(mockOctokit.gists.create).toHaveBeenCalledWith({
      description: 'Electron Fiddle Gist',
      files: {
        'index.html': { content: 'html-content' },
        'renderer.js': { content: 'renderer-content' },
        'main.js': { content: 'main-content' },
      },
      public: true
    });
  });

  it('handles an error in Gist publishing', async () => {
    const mockOctokit = {
      authenticate: jest.fn(),
      gists: {
        create: jest.fn(() => {
          throw new Error('bwap bwap');
        })
      }
    };

    (getOctokit as any).mockReturnValue(mockOctokit);

    const wrapper = shallow(<PublishButton appState={this.store} />);
    const instance: PublishButton = wrapper.instance() as any;

    await instance.publishFiddle();

    expect(mockOctokit.authenticate).toHaveBeenCalled();
    expect(wrapper.state('isPublishing')).toBe(false);
  });
});

