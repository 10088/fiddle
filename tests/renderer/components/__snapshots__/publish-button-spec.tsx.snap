// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Publish button component renders 1`] = `
<button
  className="button button-publish"
  disabled={false}
  onClick={[Function]}
>
  <FontAwesomeIcon
    border={false}
    className=""
    fixedWidth={false}
    flip={null}
    icon={
      Object {
        "icon": Array [
          512,
          512,
          Array [],
          "f093",
          "M296 384h-80c-13.3 0-24-10.7-24-24V192h-87.7c-17.8 0-26.7-21.5-14.1-34.1L242.3 5.7c7.5-7.5 19.8-7.5 27.3 0l152.2 152.2c12.6 12.6 3.7 34.1-14.1 34.1H320v168c0 13.3-10.7 24-24 24zm216-8v112c0 13.3-10.7 24-24 24H24c-13.3 0-24-10.7-24-24V376c0-13.3 10.7-24 24-24h136v8c0 30.9 25.1 56 56 56h80c30.9 0 56-25.1 56-56v-8h136c13.3 0 24 10.7 24 24zm-124 88c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20zm64 0c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20z",
        ],
        "iconName": "upload",
        "prefix": "fas",
      }
    }
    inverse={false}
    listItem={false}
    mask={null}
    pull={null}
    pulse={false}
    rotation={null}
    size={null}
    spin={false}
    symbol={false}
    transform={null}
  />
  <span
    style={
      Object {
        "marginLeft": 8,
      }
    }
  />
  Publish
</button>
`;
